release-19.2: colexec: skip over unneeded columns with unknown types in cfetcher

asubiotto opened the pull request:
Previously, we would return an error if we encountered an unknown type in the
cfetcher, even if the values from the column were never decoded into the output
coldata.Batch. This change supports reading from a table with unknown types as
long as those columns are marked as not needed. The approach this patch takes
is to allow a coldata.Batch to be created with unknown types, but disallowing
any uses of unknown type Vecs, other than observing that it is of unknown type.

Release note (sql change): Vectorized queries that execute only on supported
types even if those types form part of a table with unsupported types are now
run through the vectorized engine. This would previously fall back to the
row-by-row execution engine.

Commit:
colexec: skip over unneeded columns with unknown types in cfetcher

Previously, we would return an error if we encountered an unknown type in the
cfetcher, even if the values from the column were never decoded into the output
coldata.Batch. This change supports reading from a table with unknown types as
long as those columns are marked as not needed. The approach this patch takes
is to allow a coldata.Batch to be created with unknown types, but disallowing
any uses of unknown type Vecs, other than observing that it is of unknown type.

Release note (sql change): Vectorized queries that execute only on supported
types even if those types form part of a table with unsupported types are now
run through the vectorized engine. This would previously fall back to the
row-by-row execution engine.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42934)
<!-- Reviewable:end -->

yuzefovich commented:
[nit]: `release-19.2` is missing from the title of the PR.

Also, there is a related test failure (https://github.com/cockroachdb/cockroach/issues/37815#issuecomment-561631592), so I would hold off on backporting until we figure out the cause of that issue. I'll dig in today.

yuzefovich committed:
colexec: fix handling of Unhandled type in cfetcher, ResetInternalBatch

We recently merged a change to allow for 'unknown' vectors to be present
among colvecs. However, we forgot to update the resetting of the batch
- unknown columns do not have nulls and don't need to be reset. Now this
is fixed.

Also this commit removes an unused field in cfetcher.

Release note: None

asubiotto commented:
Thanks, will wait for that to merge and add it to this PR

asubiotto commented:
Backporting 656b636236390827bd6b8a9f8cd6b7bb4a9e2add as well. Will wait for #43002 because I think we probably need that as well.

asubiotto commented:
Actually, we should backport this first and then @rohany's changes

rohany commented:
LGTM

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 8 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42934#-:-LvSuPrDDSKDQas-37AT:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rohany)



<!-- Sent from Reviewable.io -->
